### PR TITLE
Update to sdk3.0.0 node 2.3.0  nonfungible token contracts

### PR DESCRIPTION
**Pull request by alekseytsekov**

### CODE REVIEW

File: libraries/NonFungibleToken/contracts/math.aes [hunk 1]
```diff
@@ -0,0 +1,32 @@
+namespace SafeMath =
```
Comment (hanssv) [inline libraries/NonFungibleToken/contracts/math.aes]: Most of this is built into the VM, not sure what is the point of this library?

File: libraries/NonFungibleToken/contracts/non-fungible-metadata-token.aes [hunk 2]
```diff
@@ -1,118 +1,106 @@
...
-  public stateful function setApprovalForAll(to : address, approved : bool) : bool =
-    ensureAllowed(Call.caller)
-    put(state{operatorApprovals[Call.caller][to] = approved})
-    true
+    clear_approval(from, token_id)
+    remove_token_from(from, token_id)
+    add_token_to(to, token_id)
+
+  public stateful function set_approval_for_all(to: address, approved: bool) =
+    put(state{m_operator_approvals[Call.caller = {}][to] = approved})
```
Comment (hanssv) [inline libraries/NonFungibleToken/contracts/non-fungible-metadata-token.aes]: 👍

File: libraries/NonFungibleToken/test/constants/error-messages.json [hunk 3]
```diff
@@ -5,5 +5,6 @@
     "LESS_TOKENS_THAN_ACCOUNT_BALANCE": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAKkJ1cm5lZCBhbW91bnQgaXMgbGVzcyB0aGFuIGFjY291bnQgYmFsYW5jZQAAAAAAAAAAAAAAAAAAAAAAAAAAAABC/9eP",
     "CANNOT_OVERRIDE_TOKEN": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGFRva2VuIGNhbid0IGJlIG92ZXJyaWRlZAAAAAAAAAAAOGD",
     "ONLY_OWNER_CAN_TRANSFER": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAHk9ubHkgdG9rZW4gb3duZXIgY2FuIHRyYW5zZmVyIQAAyRj",
-    "NOT_AN_OWNER_OR_NOT_APPROVED": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGU5vdCBvd25lciBvciBub3QgYXBwcm92ZWQAAAAAAAAAJTO5dA"
+    "NOT_AN_OWNER_OR_NOT_APPROVED": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGU5vdCBvd25lciBvciBub3QgYXBwcm92ZWQAAAAAAAAAJTO5dA",
```
Comment (hanssv) [inline libraries/NonFungibleToken/test/constants/error-messages.json]: I guess just decoding them is far too expensive...

File: libraries/NonFungibleToken/contracts/math.aes (hunk 1 above)
Comment (hanssv) [reply libraries/NonFungibleToken/contracts/math.aes]: ~~Most~~ All(?)

File: libraries/NonFungibleToken/contracts/non-fungible-full-token.aes [hunk 4]
```diff
@@ -1,145 +1,115 @@
...
-    tokenOwner        : map(token_id, address),
-    tokenApprovals    : map(token_id, address),
-    operatorApprovals : map(address, map(address, bool)),
-    tokenURIs         : map(token_id, string)}
-
-  public stateful function  init(name : string, symbol : string) = {
+    name: string,
+    symbol: string,
+    owner: address,
+    m_owned_tokens_count: map(address, int),
```
Comment (mradkov) [inline libraries/NonFungibleToken/contracts/non-fungible-full-token.aes]: Is there any particular reason these are prefixed with `m_`?

File: libraries/NonFungibleToken/contracts/non-fungible.aes [hunk 5]
```diff
@@ -1,110 +1,75 @@
...
-    tokenApprovals    : map(int, address),
-    operatorApprovals : map(address, map(address, bool))}
-
-  //todo repalce {} with Map.empty
-  public stateful function  init(name : string, symbol : string) = {
+    name: string,
+    symbol: string,
+    owner: address,
+    m_owned_tokens_count: map(address, int),
+    m_tokenOwner: map(int, address),
```
Comment (mradkov) [inline libraries/NonFungibleToken/contracts/non-fungible.aes]: `m_tokenOwner` should be `m_token_owner` according to code style conventions.

File: libraries/NonFungibleToken/contracts/math.aes (hunk 1 above)
Comment (mradkov) [reply libraries/NonFungibleToken/contracts/math.aes]: Agree with @hanssv

File: libraries/NonFungibleToken/contracts/non-fungible.aes [hunk 6]
```diff
@@ -1,110 +1,75 @@
...
-  public function symbol() : string = state.symbol
-  public function balanceOf(tokenOwner : address) : int = lookupByAddress(tokenOwner, state.ownedTokensCount, 0)
-  public function ownerOf(tokenId : int) = lookupById(tokenId, state.tokenOwner, #0)
-  public function getApproved(tokenId : int) : address = lookupById(tokenId, state.tokenApprovals, #0)
-  public function isApprovedForAll(owner : address, operator : address) : bool = state.operatorApprovals[owner][operator]
-
-  public stateful function approve(tokenId : int, to : address) : bool =
-    let owner : address = ownerOf(tokenId)
+    m_owned_tokens_count = {},
+    m_tokenOwner = {},
```
Comment (mradkov) [inline libraries/NonFungibleToken/contracts/non-fungible.aes]: Same like above: `m_tokenOwner` should be `m_token_owner`

File: libraries/NonFungibleToken/contracts/non-fungible.aes [hunk 7]
```diff
@@ -1,110 +1,75 @@
...
+    state.name
+
+  public function symbol() : string =
+    state.symbol
+
+  public function balance_of(token_owner: address) : int =
+    Map.lookup_default(token_owner, state.m_owned_tokens_count, 0)
+
+  public function owner_of(token_id: int) =
+    Map.lookup_default(token_id, state.m_tokenOwner, #0)
```
Comment (mradkov) [inline libraries/NonFungibleToken/contracts/non-fungible.aes]: `state.m_tokenOwner` => `state.m_token_owner`

File: libraries/NonFungibleToken/contracts/non-fungible.aes [hunk 8]
```diff
@@ -1,110 +1,75 @@
...
\ No newline at end of file
+    clear_approval(from, token_id)
+    remove_token_from(from, token_id)
+    add_token_to(to, token_id)
+
+  public stateful function set_approval_for_all(to: address, approved: bool) =
+    put(state{m_operator_approvals[Call.caller = {}][to] = approved})
+
+  private stateful function add_token_to(to: address, token_id: int) = 
+    put(state{m_tokenOwner = state.m_tokenOwner{[token_id] = to}})
```
Comment (mradkov) [inline libraries/NonFungibleToken/contracts/non-fungible.aes]: `put(state{m_tokenOwner = state.m_tokenOwner{[token_id] = to}})`

to 

`put(state{m_token_owner = state.m_token_owner{[token_id] = to}})`

File: libraries/NonFungibleToken/contracts/non-fungible.aes [hunk 9]
```diff
@@ -1,110 +1,75 @@
...
+    put(state{m_owned_tokens_count[to] = SafeMath.add(Map.lookup_default(to, state.m_owned_tokens_count, 0), 1)})
+
+  private stateful function remove_token_from(from: address, token_id: int) =
+    put(state{m_tokenOwner = state.m_tokenOwner{[token_id] = #0}})
```
Comment (mradkov) [inline libraries/NonFungibleToken/contracts/non-fungible.aes]: Change

`put(state{m_tokenOwner = state.m_tokenOwner{[token_id] = #0}})`

to 

`put(state{m_token_owner = state.m_token_owner{[token_id] = #0}})`

Comment (thepiwo) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: what is the `m_` semantic?

Comment (mradkov) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: I was asking the same here:
https://github.com/aeternity/aepp-sophia-examples/pull/37#discussion_r276681580

My comment here is to point out that snake_case should be used consistently. `m_tokenOwner` => `m_token_owner`

Comment (maptuhec) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: the **m_** prefix is used to differentiate Maps from other variables

Comment (mradkov) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: > the **m_** prefix is used to differentiate Maps from other variables

Is it necessary?

Comment (maptuhec) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: In my opinion, yes it is necessary and it is easier to read if you can make this difference

Comment (thepiwo) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: but why m_ instead if map_ or another semantic to let devs know

Comment (alekseytsekov) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: is this more readable: 
```
private function foldr(f : (('a, 'b) => 'b), z: 'b, l : list('a)) : 'b =
    switch(l)
      [] => z
      e :: l' => f(e, foldr(f, z, l'))
```
what is `z`? 
or should we type full name always like f => `function'` and l => `list`

Comment (maptuhec) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: @thepiwo @mradkov we thought the **m_** prefix is descriptive enough, and if we use **map** will make the variables too long.

Comment (mradkov) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: > @thepiwo @mradkov we thought the **m_** prefix is descriptive enough, and if we use **map** will make the variables too long.

I thought it was used for this after looking over the code in the first place, but in Solidity, for example, nobody is using something similar when naming variables and properties.

I think using something like this, here in the Sophia examples, we are pretty much influencing the language convention that should be used by developers when developing on aeternity.

That being said - in my opinion using the longer `map_` prefix shouldn't be a problem - it doesn't affect price, fees or something else. So if we proceed with the prefix at all - I'm vouching for `map_`.

 After all the code should be easily readable for humans. :)

Comment (thepiwo) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: `z` is the starting value, the initial value of the accumulator

Comment (thepiwo) [reply libraries/NonFungibleToken/contracts/non-fungible.aes]: `list` and `function` are reserved keywords, I guess they could be used, but for library functions that represent no business logic I did choose to keep them as short as possible

**Comment by thepiwo** [comment]
@alekseytsekov should we merge?

**Pull request merged**